provider/service.py

```diff
@@ -13,62 +13,115 @@
+                            logging.info('[{}] Found a new trigger to create'.format(change["id"]))
                             self.createAndRunConsumer(document)
-                        elif existingConsumer.desiredState() == Consumer.State.Running and not self.__isTriggerDocActive(document):
-                            # running trigger should become disabled
-                            logging.info('[{}] Existing running trigger should become disabled'.format(change["id"]))
-                            existingConsumer.disable()
                         else:
-                            logging.debug('[changes] Found non-interesting trigger change: \n{}\n{}'.format(existingConsumer.desiredState(), document))
-                else:
-                    logging.debug('[changes] Found a change for a non-trigger document')
+                            logging.info('[{}] Found a change to an existing trigger'.format(change["id"]))
+                            existingConsumer = self.consumers.getConsumerForTrigger(change["id"])
 
-            logging.error("[changes] uh-oh! I made it out of the changes for loop!")
+                            if existingConsumer.desiredState() == Consumer.State.Disabled and self.__isTriggerDocActive(document):
+                                # disabled trigger has become active
+                                logging.info('[{}] Existing disabled trigger should become active'.format(change["id"]))
+                                self.createAndRunConsumer(document)
+                            elif existingConsumer.desiredState() == Consumer.State.Running and not self.__isTriggerDocActive(document):
+                                # running trigger should become disabled
+                                logging.info('[{}] Existing running trigger should become disabled'.format(change["id"]))
+                                existingConsumer.disable()
+                            else:
+                                logging.debug('[changes] Found non-interesting trigger change: \n{}\n{}'.format(existingConsumer.desiredState(), document))
+                    elif 'canary' in change['doc']:
+                        # found a canary - update lastCanaryTime
+                        logging.info('[canary] I found a canary. The last one was {} seconds ago.'.format(secondsSince(self.lastCanaryTime)))
+                        self.lastCanaryTime = datetime.now()
+
+                        # delete the canary document
+                        self.database.deleteDoc(change['id'])
+                    else:
+                        logging.debug('[changes] Found a change for a non-trigger document')
 
+                if secondsSince(self.lastCanaryTime) > canaryTimeout:
+                    logging.warn('[canary] It has been more than {} seconds since the last canary - restarting the DB changes feed'.format(canaryTimeout))
+                    self.restartChangesFeed()
+                    break
```

Comment:
This break is slightly confusing since the code block is so big. At first I thought you were jumping out of the main loop. Had to scroll up to the top of the function to see what was going on. Looks like you're just skipping the log message below.